themes/classic/templates/_partials/microdata/product-jsonld.tpl

```diff
@@ -0,0 +1,107 @@
+      "@type": "Thing",
+      "name": "{if $product_manufacturer->name}{$product_manufacturer->name|escape:'html':'UTF-8'}{else}{$shop.name}{/if}"
+    },{/if}
+    {if isset($nbComments) && $nbComments && $ratings.avg}"aggregateRating": {
+      "@type": "AggregateRating",
+      "ratingValue": "{$ratings.avg|round:1|escape:'html':'UTF-8'}",
+      "reviewCount": "{$nbComments|escape:'html':'UTF-8'}"
+    },{/if}
+    {if isset($product.weight) && ($product.weight != 0)}
+      "weight": {
+        "@context": "https://schema.org",
+        "@type": "QuantitativeValue",
+        "value": "{$product.weight}",
+        "unitCode": "{$product.weight_unit}"
+      },{/if}
+      {if $product.show_price}
+        {if $product.id_product_attribute == 0}
+          "offers": {
+            "@type": "Offer",
+            "priceCurrency": "{$currency.iso_code}",
+            "name": "{$product.name|strip_tags:false}",
+            "price": "{$product.price_amount}",
+            "url": "{$product.url}",
+            "priceValidUntil": "{($smarty.now + (int) (60*60*24*15))|date_format:"%Y-%m-%d"}",
+            {if $product.images|count > 0}
+              "image": {strip}[
+                {foreach from=$product.images item=p_img name="p_img_list"}
+                  "{$p_img.large.url}"{if not $smarty.foreach.p_img_list.last},{/if}
+                {/foreach}
+              ]{/strip},
+            {/if}
+            "sku": "{if $product.reference}{$product.reference}{else}{$product.id}{/if}",
+            "mpn": "{if $product.mpn}{$product.mpn}{elseif $product.reference}{$product.reference}{else}{$product.id}{/if}",
+            {if $product.condition == 'new'}"itemCondition": "https://schema.org/NewCondition",{/if}
+            {if $product.show_condition > 0}
+              {if $product.condition == 'used'}"itemCondition": "https://schema.org/UsedCondition",{/if}
+              {if $product.condition == 'refurbished'}"itemCondition": "https://schema.org/RefurbishedCondition",{/if}
+            {/if}
+            "availability": "{if $product.quantity > 0 || $product.allow_oosp > 0}https://schema.org/InStock{else}https://schema.org/OutOfStock{/if}",
+            "seller": {
+              "@type": "Organization",
+              "name": "{$shop.name}"
+            }
+          }
+        {else}
+          "offers": [
+              {foreach key=id_product_combination item=combination from=$combinations}
+                {
+                  "@type": "Offer",
+                  "priceCurrency": "{$currency.iso_code}",
+                  "name": "{$product.name|strip_tags:false} - {foreach item=attr from=$combination.attributes_values}{$attr|strip_tags:false}
+                    {if !$attr@last} 
+                    {/if}
+                  {/foreach}",
+                  "price": "{Product::getPriceStatic($product->id, true, $id_product_combination)|round:'2'}",
+                  "url": "{$product.url}",
```

Comment:
Can we add URL to the specific combination here?

Reply:
I think not, if we have different urls for combinations we don't have combinations defined as offers for the product. It's a different product for Google... (avoid duplications)
If you have the same url for all product combinations and if you don't define combinations as offer microdata Google will miss your (colors, flavors etc).
I don't know if it's clear but (different urls => other product for Google / same url  => same product but with variant defined as Offer)

Reply:
I think this part and the conditions need to be tested a lot in different scenarios (shop configuration) and the JSONLD declaration need to be adapted accordingly.
The configuration I m sure it works great with is : firendly url without combination id in the url => 1 cannonical url per product (aligned with Google recommendation).
For the rest I will need help to test and adapt if needed. 
The assumption is : one page per product (combinations as offer) / if one product page per combination at the end the logic is to create a completlly different product ?

Reply:
@Hlavtox more I think about "what Goolge really wants" more I m in the fog :) For sure, if someone is clear in his mind about the "best way", i will adapt.

Reply:
@Hlavtox I think my "combinations as Offer" proposition is not aligned with a normal shop config (on my side I rewrite to eliminate the product_id and the combination_id from the url).
do you know how I can check here
line 54 of product-jsonld.tpl `{if $product.id_product_attribute == 0}`

I would like to replace by something like this 
`{if $product.id_product_attribute == 0 || IsCombination id is in the url}` 
:)

I think it could depend of the sitemap too :( (I don't remember if the product combinations have been set with noindex) => in this case, your proposition to have the combination url is good. (One Indexed url per product, but url information in the offer for combinations).

Reply:
at the end @hlavtox, I think your propostion to have the combination url in the $combination smarty var can solve at least one part of the issue and it can be fine for the moment. 

Because the general issue is more related to the Presta Core. IMO we have 3 ways 

**1) Not very SEO oriented**

- Sitemap transfered to Goolge doesn't exclude combinations URLs
- No URL rewriting (1 url per combination)
- See if we can declare the same list of Offers multiples times with the correct combination url or limit to one offer per page(combination).
- It will not solve that Google will consider some content as duplicate (I think)

**2) Sitemap exclude combinations pages**

- Sitemap transfered to Goolge exclude combinations url with noindex
- No URL rewriting (1 url per combination)
- In this case, declare several offers per product => one offer for each combination (if possible with the correct url of the combination 

**3) Url rewriting (no product id, no combination id in url)**

- IMO the best approach, but not Presta standard
- Sitemap sends only one url per product (no duplicate)
- One product and several offers for combinations
- The clean way (my PR maybe take to much that into account)

Summary : maybe we need to completly remove the combinations part in product-jsonld.tpl to remain standard but i will be sad :)

Reply:
See above -> open point on definition of product-jsonld.tpl